wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/command/CommandScheduler.java

```diff
@@ -0,0 +1,352 @@
+
+    if (m_disabled
+        || (RobotState.isDisabled() && !command.getRunWhenDisabled())
+        || m_scheduledCommands.containsKey(command)) {
+      return;
+    }
+
+    Set<Subsystem> requirements = command.getRequirements();
+
+    if (Collections.disjoint(m_requirements.keySet(), requirements)) {
+      command.initialize();
+      CommandState scheduledCommand = new CommandState(interruptible);
+      m_scheduledCommands.put(command, scheduledCommand);
+      for (Consumer<Command> action : m_initActions) {
+        action.accept(command);
+      }
+      for (Subsystem requirement : requirements) {
+        m_requirements.put(requirement, command);
+      }
+    } else {
+      boolean allInterruptible = true;
+      for (Subsystem requirement : requirements) {
+        if (m_requirements.keySet().contains(requirement)) {
+          allInterruptible &=
+              m_scheduledCommands.get(m_requirements.get(requirement)).isInterruptible();
+        }
+      }
+      if (allInterruptible) {
+        for (Subsystem requirement : requirements) {
+          if (m_requirements.containsKey(requirement)) {
+            cancelCommand(m_requirements.get(requirement));
+          }
+        }
+        command.initialize();
+        CommandState scheduledCommand = new CommandState(interruptible);
+        m_scheduledCommands.put(command, scheduledCommand);
+        for (Consumer<Command> action : m_initActions) {
+          action.accept(command);
+        }
+      }
+    }
+  }
+
+  /**
+   * Runs a single iteration of the scheduler.  The execution occurs in the following order:
+   *
+   * <p>Subsystem periodic methods are called.
+   *
+   * <p>Button bindings are polled, and new commands are scheduled from them.
+   *
+   * <p>Currently-scheduled commands are executed.
+   *
+   * <p>End conditions are checked on currently-scheduled commands, and commands that are finished
+   * have their end methods called and are removed.
+   *
+   * <p>Any subsystems not being used as requirements have their default methods started.
+   */
+  @SuppressWarnings({"PMD.CyclomaticComplexity", "PMD.NPathComplexity"})
+  public void run() {
+
+    if (m_disabled) {
+      return;
+    }
+
+    for (Subsystem subsystem : m_subsystems) {
+      subsystem.periodic();
+    }
+
+    for (Trigger.ButtonScheduler button : m_buttons) {
+      button.execute();
+    }
+
+    for (Command command : m_scheduledCommands.keySet()) {
+
+      if (RobotState.isDisabled() && !command.getRunWhenDisabled()) {
+        cancelCommand(command);
+        continue;
+      }
+
+      command.execute();
+      for (Consumer<Command> action : m_executeActions) {
+        action.accept(command);
+      }
+      if (command.isFinished()) {
+        command.end();
+        for (Consumer<Command> action : m_endActions) {
+          action.accept(command);
+        }
+        m_scheduledCommands.remove(command);
+        m_requirements.keySet().removeAll(command.getRequirements());
```

Comment:
I know its hard, but we should try and make our built in commands compute requirements during initialization, and return a constant set instance to avoid all these allocations. Hard to do that for use implementations because of the interface, but worth doing for our implementations.

Reply:
Fair point.  It's not that hard to do, and we can advise teams in the documentation to do the same.

Reply:
I've now done this for the two CommandGroup classes.